Explainer for proposed stream changes related to VideoFrame processing (#1193)
whatwg/streams: streams-for-raw-video-explainer.md

By doing so, the processing unit that enqueues/writes objects will not be able to mutate objects manipulated by the stream and is relieved of the lifetime management of these objects.
Conversely, processing units take ownership of objects when they receive them from a stream.

Transferring ownership should be opt-in. For that purpose, a new streams type, named 'transfer' in this document,  would be added.

## Example

Below is an example of JavaScript that shows how this can be used.
The example creates a processing pipe starting with a camera stream and applying two transforms, one for doing a processing for every 30 frame, and one for doing background blur.

```javascript
// JS transform
const frameCountTransform = new TransformStream({
  transform: async (videoFrame, controller) => {
    try {
      // videoFrame is under the responsibility of the script and must be closed when no longer needed
      controller.enqueue(videoFrame);
      // At this point, videoFrame has been transfered within controller.enqueue call. frameCountTransform cannot mutate it.
      if (!controller.count)

>         controller.count = 0;

Reviewer: It might be a bit confusing to use "expando properties" in these examples. (What if we decide to add `TransformStreamDefaultController.prototype.count` in the future? 😛)

For these examples, it's probably better to use a regular variable:
```javascript
let count = 0;
const frameCountTransform = new TransformStream({
  transform: async (videoFrame, controller) => { /* ... */ }
});
```

Author: Will update the example

Author: Done

>       if (!(i++controller.count % 30) && frameCountTransform.onEach30Frame)

Reviewer: `i++controller.count` is a `SyntaxError`. 😬

```suggestion
      if (!(++controller.count % 30) && frameCountTransform.onEach30Frame)
```

const backgroundBlurTransform = new BackgroundBlurTransform();

const cameraStream = await navigator.mediaDevices.getUserMedia({ video : true });
const videoFrameStream = getReadableStreamFromTrack(cameraStream.getVideoTracks()[0]);
const blurredVideoFrameStream = videoFrameStream.pipeThrough(backgroundBlurTransform)
                                                .pipeThrough(frameCountTransform);
const blurredStream = new MediaStream([getTrackFromReadableStream(blurredVideoFrameStream)]);
// Make use of blurredStream.
...
```

## Goals

*   Permit `ReadableStream`, `WritableStream` and `TransformStream` objects to take ownership of objects they manipulate.
*   Permit to build a safe and optimal video pipeline using `ReadableStream`, `WritableStream` and `TransformStream` objects that manipulate `VideoFrame` objects.
*   Permit both native and JavaScript-based streams of type 'transfer'.
*   Permit to optimize streams pipelines of transferable objects like `ArrayBuffer`, `RTCEncodedVideoFrame` or `RTCEncodedAudioFrame`.
*   Permit to tee a `ReadableStream` of `VideoFrame` objects without tight coupling between the teed branches.

## Non-goals

> *   Add support for supporting transfer and closing of arbitrary JavaScript objects.

Reviewer: ```suggestion
*   Add support for transferring and closing of arbitrary JavaScript objects.
```

## Use cases

*   Performing realtime transformations of `VideoFrame` objects, for instance taking a camera `MediaStreamTrack` and applying
    a background blur effect as a `TransformStream` on each `VideoFrame` of the `MediaStreamTrack`.

## End-user benefit

*   `VideoFrame` needs specific management and be closed as quickly as possible, without relying on garbage collection.
    This is important to not create hangs/stutters in the processing pipeline. By building support for safe patterns
    directly in streams, this will allow web developers to optimize `VideoFrame` management, and allow user experience
    to be more consistent accross devices.

## Principles

The envisioned changes to the streams specification could look like the following:
*   Add a new 'transfer' type to `ReadableStream`, `WritableStream` and `TransformStream`.

>     For streams that do not have the 'transfer' type, nothing changes.

Reviewer: A `TransformStream` may transform one type of chunks to a different type of chunks. For example, a `TextDecoderStream` accepts `ByteSource`s on its writable end and produces `string`s on its readable end. We should try to support such use cases for ownership-transferring streams too. Therefore, we shouldn't "just" have `type: "transfer"` for `TransformStream`.

We have already reserved `readableType` and `writableType` for such use cases, see the [`TransformStream` constructor](https://streams.spec.whatwg.org/#ts-constructor). I suggest we use that:
* If `writableType` is `"transfer"`, then the writable end will accept transferable objects. Ownership is transferred into the stream when calling `writer.write(chunk)`, and transferred out when the stream calls `transformer.transform(chunk, controller)`.
* If `readableType` is `"transfer"`, then the readable end will produce transferable objects. Ownership is transferred into the stream when calling `controller.enqueue(chunk)`, and transferred out when resolving a `reader.reader()` call.

For convenience, we *could* still add `transformer.type`, as a shorthand for setting both `readableType` and `writableType` to the same value. That would make the common case of transforming a `VideoFrame` to another `VideoFrame` easier. Or maybe we want to leave that as a future improvement.

Author: Right, I'll update this line for TransformStream.

Author: Done

> *   Streams of the 'transfer' type can only manipulate Transferable or Serializable objects.

Reviewer: I believe we have to *require* all objects to be `Serializable`. Otherwise, we cannot call `StructuredSerialize`/`StructuredDeserialize` inside `tee()` to make a "real" clone. (A transferred version of an object is not a clone, since you can no longer use the original.)

Author: The two options are only support Serialisable, or throw.
The use case only requires Serializable right now, so let's go with this.

Author: Restricted to Transferable+Serializable for now.
Added open questions for only transferable and only serializable.

>     If a non Transferable or Serializable object is enqueued or written, the object is ignored as if it was never enqueued/written.

Reviewer: Hmm, not sure if I like this "silent error". For readable byte streams, we throw a `TypeError` in `controller.enqueue()` if the given chunk is not an `ArrayBufferView`. Ideally, we'd do something similar for ownership-transferring streams.

Author: I'll add an open question section to mention what to do when enqueuing/writing not serializable objects.
I prefer rejecting over being silent if we are ok not erroring the stream either.

Author: Added

Reviewer: Agreed. If we can make this an early error that doesn't error the entire stream, that'd be great. 🙂

*   If a Transferable object is enqueueud/written in a 'transfer' type `ReadableStreamDefaultController`, `TransformStreamDefaultController`

>     or `WritableStreamDefaultWriter`, create a copy of the object using StructuredSerializeWithTransfer/StructuredDeserializeWithTransfer.
>     Proceed with the regular stream algorithm by using the copy of the object instead of the object itself.

Reviewer: ```suggestion
    or `WritableStreamDefaultWriter`, create a transferred version of the object using StructuredSerializeWithTransfer/StructuredDeserializeWithTransfer.
    Proceed with the regular stream algorithm by using the transferred object instead of the object itself.
```

*   If a Serializable object is enqueueud/written in a 'transfer' type `ReadableStreamDefaultController`, `TransformStreamDefaultController`
    or `WritableStreamDefaultWriter`, create a copy of the object using StructuredSerialize/StructuredDeserialize.

>     Proceed with the regular stream algorithm by using the copy of the object instead of the object itself.

Reviewer: We should also close the original object if it supports `Closeable`. (If it doesn't support it, we just let it be garbage collected as normal.)

Author: This is probably safer but probably unneeded in the case of transferable (& serialisable) since the object would have been transferred then closed. I guess the question remains for serialisable but not transferable objects. Use case only requires Transferable+Serializable.

Author: Restricted to Transferable+Serializable and making use of transfer steps.

Author: Plus added an item about serializable only objects

*   Introduce a WhatWG streams 'close-able' concept. An object that is 'close-able' defines closing steps.
    For instance `VideoFrame` closing steps could be defined using https://www.w3.org/TR/webcodecs/#close-videoframe.

>     `ArrayBuffer` closing steps could be defined using https://tc39.es/ecma262/#sec-detacharraybuffer.

Reviewer: `DetachArrayBuffer` doesn't deallocate the underlying Data Block of the `ArrayBuffer`, since it's also used to *transfer* an `ArrayBuffer` (see [step 5.4 of `StructuredSerializeWithTransfer`](https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializewithtransfer)). However, in our case we know that we won't transfer it to a new `ArrayBuffer`, and thus the Data Block *could* be deallocated immediately.

We should make a note that implementations are *allowed* to deallocate the Data Block immediately after closing an `ArrayBuffer` (or they could just wait for garbage collection as normal).

Author: Good point, not sure that it fits the explainer but would definitely be good for a PR.

Author: Added a corresponding item in future work.

        if (!(++frameCount % 30) && doLogging)
            doLogging(frameCount);
      } catch (e) {
        // In case of exception, let's make sure videoFrame is closed. This is a no-op if videoFrame was previously transferred.
        videoFrame.close();
        // If exception is unrecoverable, let's error the pipe.
        controller.error(e);
      }
    },
    readableType: 'transfer',
    writableType: 'transfer'
  });
  // Native transform is of type 'transfer'
  const backgroundBlurTransform = new BackgroundBlurTransform();

  return videoFrameStream.pipeThrough(backgroundBlurTransform)
                         .pipeThrough(frameCountTransform);
}
```

## Goals

*   Permit `ReadableStream`, `WritableStream` and `TransformStream` objects to take ownership of chunks they manipulate.
*   Permit to build a safe and optimal video pipeline using `ReadableStream`, `WritableStream` and `TransformStream` objects that manipulate `VideoFrame` objects.
*   Permit both native and JavaScript-based streams of type 'transfer'.
*   Permit to optimize streams pipelines of transferable chunks like `ArrayBuffer`, `RTCEncodedVideoFrame` or `RTCEncodedAudioFrame`.

> *   Permit to tee a `ReadableStream` of `VideoFrame` objectss without tight coupling between the teed branches.

Reviewer: ```suggestion
*   Permit to tee a `ReadableStream` of `VideoFrame` objects without tight coupling between the teed branches.
```

If A does not close `VideoFrame`, it is the responsibility of the remaining of the pipeline to close it.
There is a need to clearly identify who is owning these objects and who is responsible to close these objects at any point in time.

The proposed solution is to transfer ownership of an object to the stream when it gets written or enqueueud to the stream.
By doing so, the processing unit that enqueues/writes objects will not be able to mutate objects manipulated by the stream and is relieved of the lifetime management of these objects.
Conversely, processing units take ownership of objects when they receive them from a stream.

Transferring ownership should be opt-in. For that purpose, a new streams type, named 'transfer' in this document,  would be added.

## Example

Below is an example of JavaScript that shows how this can be used.
The example creates a processing pipe starting with a camera stream and applying two transforms, one for doing a processing for every 30 frame, and one for doing background blur.

```javascript

> // JS transform

Reviewer 2: I think it's important we lead with worker examples from the get-go.
```suggestion
// worker.js
```

Author: Right

let frameCount = 0;
const frameCountTransform = new TransformStream({
  transform: async (videoFrame, controller) => {
    try {
      // videoFrame is under the responsibility of the script and must be closed when no longer needed
      controller.enqueue(videoFrame);
      // At this point, videoFrame has been transfered within controller.enqueue call. frameCountTransform cannot mutate it.
      if (!(++frameCount % 30) && frameCountTransform.onEach30Frame)
          frameCountTransform.onEach30Frame(frameCount);
    } catch {
      videoFrame.close();

>     }

Author: > It cannot mutate it, because it cannot even access it, is that right?

It has access to the object, but no longer to the underlying bytes.

> Also, don't we need to rethrow in the catch, to not encourage overriding default error handling?

One of the complaint we heard from streams was that errors propagate very easily, while there may be recoverable errors hence why I did not add throw there. I would prefer calling the error method explicitly in case of unrecoverable errors.

> I also worry this risks calling close on a transferred object, or is that benign? 

I think it is begnin.

>     }

Reviewer 3: If we borrow the "transfer" concept from "transferable", the object left behind after the transfer is "neutered"; usually this means all the methods on it check a "neutered" flag before doing anything. (I may have the name wrong.)

I think we should mandate that all closable objects have a close operation that is a no-op if the object is already closed (or transferred).

>     }

Author: Right, the explainer refers to https://www.w3.org/TR/webcodecs/#close-videoframe which will be a no-op in practice if you call it more than once.

>     }

Author: > I think we should mandate that all closable objects have a close operation that is a no-op if the object is already closed (or transferred).

I added a line to explicitly state this.

  },
  readableType: 'transfer',
  writableType: 'transfer'
});
frameCountTransform.onEach30Frame = (frameCount) => {
  // Do something periodically.
};
// Native transform is of type 'transfer'
const backgroundBlurTransform = new BackgroundBlurTransform();

> const cameraStream = await navigator.mediaDevices.getUserMedia({ video : true });
> const videoFrameStream = getReadableStreamFromTrack(cameraStream.getVideoTracks()[0]);
> const blurredVideoFrameStream = videoFrameStream.pipeThrough(backgroundBlurTransform)
>                                                 .pipeThrough(frameCountTransform);
> const blurredStream = new MediaStream([getTrackFromReadableStream(blurredVideoFrameStream)]);
> // Make use of blurredStream.

Reviewer 2: Use workers part 2. Also, mediacapture-transform is (almost as of this writing) adopted, so I think we can be specific now:
```suggestion
self.onmessage = async ({data: {cameraTrack}}) => {
  const {readable} = new MediaStreamTrackProcessor({cameraTrack});
  const {writable, track} = new VideoTrackGenerator();
  parent.postMessage({track}, [track]);
  await readable.pipeThrough(backgroundBlurTransform).pipeThrough(frameCountTransform).pipeTo(writable);
}

// main.js
const [cameraTrack] = (await navigator.mediaDevices.getUserMedia({video: true}).getVideoTracks();
const worker = new Worker('worker.js');
worker.postMessage({cameraTrack}, [cameraTrack]);

const {data} = await new Promise(r => worker.onmessage);
const blurredStream = new MediaStream([data.track]);
video.srcObject = blurredStream;
```

Author: I prefer we keep it abstract from MediaStreamTrackProcessor and we keep the example as small as possible.
I'll remove the call to getUserMedia which is indeed confusing.

*   Permit to tee a `ReadableStream` of `VideoFrame` objects without tight coupling between the teed branches.

## Non-goals

*   Add support for transferring and closing of arbitrary JavaScript objects.

## Use cases

*   Performing realtime transformations of `VideoFrame` objects, for instance taking a camera `MediaStreamTrack` and applying
    a background blur effect as a `TransformStream` on each `VideoFrame` of the `MediaStreamTrack`.

## End-user benefit

*   `VideoFrame` needs specific management and be closed as quickly as possible, without relying on garbage collection.
    This is important to not create hangs/stutters in the processing pipeline. By building support for safe patterns
    directly in streams, this will allow web developers to optimize `VideoFrame` management, and allow user experience
    to be more consistent accross devices.

## Principles

The envisioned changes to the streams specification could look like the following:
*   Add a new 'transfer' value that can be passed to `ReadableStream` type, `WritableStream` type and `TransformStream` readableType/writableType.
    For streams that do not use the 'transfer' type, nothing changes.

> *   Streams of the 'transfer' type can only manipulate that are marked both as Transferable and Serializable.

Author: Moving to chunk is probably more consistent.

> *   Streams of the 'transfer' type can only manipulate that are marked both as Transferable and Serializable.

Author: Moved to chunk

*   If an object that is either not Transferable or not Serializable is enqueued or written, the object is ignored as if it was never enqueued/written.
*   If a Transferable and Serializable object is enqueueud/written in a 'transfer' type `ReadableStreamDefaultController`, `TransformStreamDefaultController`
    or `WritableStreamDefaultWriter`, create a transferred version of the object using StructuredSerializeWithTransfer/StructuredDeserializeWithTransfer.
    Proceed with the regular stream algorithm by using the transferred object instead of the object itself.

> *   Introduce a WhatWG streams 'close-able' concept. An object that is 'close-able' defines closing steps.

Author: The idea is to start with a WhatWG stream closable or closable concept, but look at making it a WebIDL concept if there is enough interest. See one of the last item in proposal

> *   Introduce a WhatWG streams 'close-able' concept. An object that is 'close-able' defines closing steps.

Author: I kept close-able for now but am happy to change to closable if found better.
Interestingly, Java defines closeable as: 'a source or destination of data that can be closed. The close method is invoked to release resources that the object is holding (such as open files).'
